src/main/java/com/google/step/YOUR_PROJECT_NAME_HERE/external/StackOverflowClient.java

```diff
@@ -0,0 +1,129 @@
+package com.google.step.YOUR_PROJECT_NAME_HERE.external;
+
+import com.google.step.YOUR_PROJECT_NAME_HERE.data.Card;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.*;
+import java.util.*;
+import org.apache.http.HttpEntity;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.select.Elements;
+
+public final class StackOverflowClient {
+  private static final String SEARCH_URL_TEMPLATE =
+      "https://api.stackexchange.com/2.2/questions/%s?"
+          + "order=desc&sort=activity&site=stackoverflow";
+  private static final String QUESTION_URL_TEMPLATE =
+      "https://api.stackexchange.com/2.2/questions/%s/answers?"
+          + "order=desc&sort=votes&site=stackoverflow";
+  // this url specify filter to generate answer body
+  private static final String ANSWER_URL_TEMPLATE =
+      "https://api.stackexchange.com/2.2/answers/%s?order"
+          + "=desc&sort=activity&site=stackoverflow&filter=!9_bDE(fI5";
+  private static final int ID_INDEX = 2;
+
+  public Card search(String so_url) {
+    try {
+      Card card = getQuestion(so_url);
+      card = getAnswerId(card);
+      card = getAnswer(card);
+      return card;
+    } catch (URISyntaxException e) {
+      return new Card();
+    }
+  }
+
+  /* Get the question id based on url from the CSE result */
+  private Card getQuestion(String so_url) throws URISyntaxException {
+    Card card = new Card();
+    URI uri = new URI(so_url);
+    // Parse the URL to get the question id.
+    String[] segments = uri.getPath().split("/");
+    String idStr = segments[ID_INDEX];
+    String url = String.format(SEARCH_URL_TEMPLATE, idStr);
+    try {
+      JSONObject res = getResponse(url);
+      String title = res.getJSONArray("items").getJSONObject(0).get("title").toString();
+      card.setLink(uri.toString());
+      // Store the id of the question in order to get the code body of the answer.
+      card.setCode(idStr);
+      card.setTitle(title);
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+    return card;
+  }
+
+  /* Get the most voted answer's id and store it in the card */
+  private Card getAnswerId(Card card) {
+    String url = String.format(QUESTION_URL_TEMPLATE, card.getCode());
+    try {
+      JSONObject res = getResponse(url);
+      JSONArray answer_list = res.getJSONArray("items");
+      String answer_id = res.getJSONArray("items").getJSONObject(0).get("answer_id").toString();
```

Comment:
Avoid nonconstant underscore variable naming here and elsewhere in the file and use camel case instead (?)

Reply:
note: also change the items to constant